File: app/src/main/java/me/rei_m/hyakuninisshu/presentation/karuta/QuizFragment.java
package me.rei_m.hyakuninisshu.presentation.karuta;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.DecelerateInterpolator;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import me.rei_m.hyakuninisshu.databinding.FragmentQuizBinding;
import me.rei_m.hyakuninisshu.presentation.BaseFragment;
import me.rei_m.hyakuninisshu.presentation.karuta.viewmodel.QuizViewModel;
import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class QuizFragment extends BaseFragment implements QuizContact.View {

    private static final int SPEED_DISPLAY_ANIMATION_MILL_SEC = 500;

    @Inject
    QuizContact.Actions presenter;

    private FragmentQuizBinding binding;

    private OnFragmentInteractionListener listener;

    private Subscription subscription;

    public QuizFragment() {
        // Required empty public constructor
    }

    public static QuizFragment newInstance() {
        return new QuizFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getComponent().inject(this);
        presenter.onCreate(this);
        if (getArguments() != null) {
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        binding = FragmentQuizBinding.inflate(inflater, container, false);
        binding.setPresenter(presenter);
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        if (subscription != null) {
            subscription.unsubscribe();
            subscription = null;
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            listener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

    @Override
    public void initialize(QuizViewModel viewModel) {
        binding.setViewModel(viewModel);

        final List<TextView> firstLine = new ArrayList<>(Arrays.asList(
                binding.phrase11,
                binding.phrase12,
                binding.phrase13,
                binding.phrase14,
                binding.phrase15
        ));

        final List<TextView> secondLine = new ArrayList<>(Arrays.asList(
                binding.phrase21,
                binding.phrase22,
                binding.phrase23,
                binding.phrase24,
                binding.phrase25,
                binding.phrase26,
                binding.phrase27
        ));

        final List<TextView> thirdLine = new ArrayList<>(Arrays.asList(
                binding.phrase31,
                binding.phrase32,
                binding.phrase33,
                binding.phrase34,
                binding.phrase35
        ));

        final List<TextView> totalKarutaTextViewList = new ArrayList<>();
        final StringBuilder totalKarutaStringBuilder = new StringBuilder();

        totalKarutaStringBuilder.append(viewModel.firstPhrase);
        Observable.from(firstLine).take(viewModel.firstPhrase.length()).subscribe(totalKarutaTextViewList::add);

        totalKarutaStringBuilder.append(viewModel.secondPhrase);
        Observable.from(secondLine).take(viewModel.secondPhrase.length()).subscribe(totalKarutaTextViewList::add);

        totalKarutaStringBuilder.append(viewModel.thirdPhrase);
        Observable.from(thirdLine).take(viewModel.thirdPhrase.length()).subscribe(totalKarutaTextViewList::add);

        Observable.from(totalKarutaTextViewList).subscribe(textView -> {
            textView.setVisibility(View.INVISIBLE);
        });

        String totalKarutaString = totalKarutaStringBuilder.toString();

        subscription = Observable.interval(SPEED_DISPLAY_ANIMATION_MILL_SEC, TimeUnit.MILLISECONDS).take(totalKarutaString.length())
                .subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread())
                .subscribe(indexL -> {
                    final int index = indexL.intValue();
                    final Animation fadeIn = new AlphaAnimation(0, 1);
                    fadeIn.setInterpolator(new DecelerateInterpolator());
                    fadeIn.setDuration(SPEED_DISPLAY_ANIMATION_MILL_SEC);

                    final TextView currentPhraseView = totalKarutaTextViewList.get(index);
                    currentPhraseView.setText(totalKarutaString.substring(index, index + 1));
                    currentPhraseView.setVisibility(View.VISIBLE);
                    currentPhraseView.startAnimation(fadeIn);
                });
    }

    @Override
    public void displayAnswer(String quizId) {
        if (listener != null) {
            listener.onAnswered(quizId);
        }
    }

    public interface OnFragmentInteractionListener {
        void onAnswered(String quizId);
    }
}

File: feature/quiz/src/main/java/me/rei_m/hyakuninisshu/feature/quiz/ui/QuizFragment.kt
/*
 * Copyright (c) 2018. Rei Matsushita
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

/* ktlint-disable package-name */
package me.rei_m.hyakuninisshu.feature.quiz.ui

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import android.view.animation.DecelerateInterpolator
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import dagger.Binds
import dagger.android.AndroidInjector
import dagger.android.support.DaggerFragment
import dagger.android.support.FragmentKey
import dagger.multibindings.IntoMap
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import me.rei_m.hyakuninisshu.databinding.FragmentQuizBinding
import me.rei_m.hyakuninisshu.feature.corecomponent.di.FragmentScope
import me.rei_m.hyakuninisshu.domain.model.quiz.KarutaQuiz
import me.rei_m.hyakuninisshu.domain.model.quiz.KarutaQuizContent
import me.rei_m.hyakuninisshu.domain.model.quiz.KarutaQuizIdentifier
import me.rei_m.hyakuninisshu.ext.withArgs
import me.rei_m.hyakuninisshu.presentation.core.di.QuizFragmentModule
import me.rei_m.hyakuninisshu.feature.corecomponent.enums.KarutaStyleFilter
import me.rei_m.hyakuninisshu.feature.corecomponent.enums.QuizAnimationSpeed
import me.rei_m.hyakuninisshu.presentation.widget.view.KarutaTextView
import me.rei_m.hyakuninisshu.feature.corecomponent.helper.AnalyticsHelper
import me.rei_m.hyakuninisshu.feature.corecomponent.flux.EventObserver
import java.util.Arrays
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class QuizFragment : DaggerFragment() {

    @Inject
    lateinit var analyticsHelper: AnalyticsHelper

    @Inject
    lateinit var viewModelFactory: QuizViewModel.Factory

    private lateinit var quizViewModel: QuizViewModel

    private lateinit var binding: FragmentQuizBinding

    private var listener: CoreInteractionListener? = null

    private val karutaQuizId by lazy {
        requireNotNull(arguments?.getParcelable<KarutaQuizIdentifier>(ARG_KARUTA_QUIZ_ID)) {
            "$ARG_KARUTA_QUIZ_ID is missing"
        }
    }

    private val kamiNoKuStyle by lazy {
        requireNotNull(arguments?.getInt(ARG_KAMI_NO_KU_STYLE)?.let { KarutaStyleFilter[it] }) {
            "$ARG_KAMI_NO_KU_STYLE is missing"
        }
    }

    private val shimoNoKuStyle by lazy {
        requireNotNull(arguments?.getInt(ARG_SHIMO_NO_KU_STYLE)?.let { KarutaStyleFilter[it] }) {
            "$ARG_SHIMO_NO_KU_STYLE is missing"
        }
    }

    private val animationSpeed by lazy {
        requireNotNull(arguments?.getInt(ARG_ANIMATION_SPEED)?.let { QuizAnimationSpeed[it] }) {
            "$ARG_ANIMATION_SPEED is missing"
        }
    }

    private var animationDisposable: Disposable? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        quizViewModel = ViewModelProviders.of(this, viewModelFactory).get(QuizViewModel::class.java)

        binding = FragmentQuizBinding.inflate(inflater, container, false).apply {
            viewModel = quizViewModel
            setLifecycleOwner(this@QuizFragment.viewLifecycleOwner)
        }

        quizViewModel.content.observe(this, Observer<KarutaQuizContent> {
            it ?: return@Observer
            when (it.quiz.state) {
                KarutaQuiz.State.IN_ANSWER -> {
                    val (firstPhrase, secondPhrase, thirdPhrase) = it.yomiFuda(kamiNoKuStyle.value)
                    startDisplayQuizAnimation(firstPhrase, secondPhrase, thirdPhrase)
                }
                KarutaQuiz.State.ANSWERED -> {
                    stopDisplayQuizAnimation()
                }
                else -> {
                }
            }
        })
        quizViewModel.openAnswerEvent.observe(this, EventObserver {
            listener?.onAnswered(it)
        })
        quizViewModel.unhandledErrorEvent.observe(this,
            EventObserver {
                listener?.onErrorQuiz()
            })

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        analyticsHelper.sendScreenView("Quiz", requireActivity())
    }

    override fun onResume() {
        super.onResume()
        quizViewModel.start()
    }

    override fun onPause() {
        stopDisplayQuizAnimation()
        super.onPause()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is CoreInteractionListener) {
            listener = context
        } else {
            throw RuntimeException("$context must implement CoreInteractionListener")
        }
    }

    override fun onDetach() {
        listener = null
        super.onDetach()
    }

    // TODO: DataBindingに寄せることができるはず.
    private fun startDisplayQuizAnimation(
        firstPhrase: String,
        secondPhrase: String,
        thirdPhrase: String
    ) {
        if (animationDisposable != null) {
            return
        }

        val animationSpeedValue = animationSpeed.value ?: return

        val firstLine = Arrays.asList<KarutaTextView>(
            binding.phrase11,
            binding.phrase12,
            binding.phrase13,
            binding.phrase14,
            binding.phrase15,
            binding.phrase16
        )
        firstLine.forEach { it.visibility = View.INVISIBLE }

        val secondLine = Arrays.asList<KarutaTextView>(
            binding.phrase21,
            binding.phrase22,
            binding.phrase23,
            binding.phrase24,
            binding.phrase25,
            binding.phrase26,
            binding.phrase27,
            binding.phrase28
        )
        secondLine.forEach { it.visibility = View.INVISIBLE }

        val thirdLine = Arrays.asList<KarutaTextView>(
            binding.phrase31,
            binding.phrase32,
            binding.phrase33,
            binding.phrase34,
            binding.phrase35,
            binding.phrase36
        )

        thirdLine.forEach { it.visibility = View.INVISIBLE }

        val totalKarutaTextViewList = ArrayList<TextView>().apply {
            addAll(firstLine.take(firstPhrase.length))
            addAll(secondLine.take(secondPhrase.length))
            addAll(thirdLine.take(thirdPhrase.length))
        }

        animationDisposable = Observable.interval(animationSpeedValue, TimeUnit.MILLISECONDS)
            .zipWith(totalKarutaTextViewList) { _, textView -> textView }
            .observeOn(AndroidSchedulers.mainThread())
            .doOnDispose {
                totalKarutaTextViewList.forEach { it.animation?.cancel() }
            }
            .subscribe { textView ->
                val fadeIn = AlphaAnimation(0f, 1f).apply {
                    interpolator = DecelerateInterpolator()
                    duration = animationSpeedValue
                }
                textView.visibility = View.VISIBLE
                textView.startAnimation(fadeIn)
            }
    }

    private fun stopDisplayQuizAnimation() {
        animationDisposable?.dispose()
        animationDisposable = null
    }

    @FragmentScope
    @dagger.Subcomponent(
        modules = [
            QuizFragmentModule::class
        ]
    )
    interface Subcomponent : AndroidInjector<QuizFragment> {
        @dagger.Subcomponent.Builder
        abstract class Builder : AndroidInjector.Builder<QuizFragment>() {

            abstract fun quizFragmentModule(module: QuizFragmentModule): Builder

            override fun seedInstance(instance: QuizFragment) {
                quizFragmentModule(
                    QuizFragmentModule(
                        instance.karutaQuizId,
                        instance.kamiNoKuStyle,
                        instance.shimoNoKuStyle
                    )
                )
            }
        }
    }

    @dagger.Module(subcomponents = [Subcomponent::class])
    abstract class Module {
        @Binds
        @IntoMap
        @FragmentKey(QuizFragment::class)
        abstract fun bind(builder: Subcomponent.Builder): AndroidInjector.Factory<out Fragment>
    }

    companion object {

        const val TAG: String = "QuizFragment"

        private const val ARG_KARUTA_QUIZ_ID = "karutaQuizId"

        private const val ARG_KAMI_NO_KU_STYLE = "kamiNoKuStyle"

        private const val ARG_SHIMO_NO_KU_STYLE = "shimoNoKuStyle"

        private const val ARG_ANIMATION_SPEED = "animationSpeed"

        fun newInstance(
            karutaQuizId: KarutaQuizIdentifier,
            kamiNoKuStyle: KarutaStyleFilter,
            shimoNoKuStyle: KarutaStyleFilter,
            animationSpeed: QuizAnimationSpeed
        ) = QuizFragment().withArgs {
            putParcelable(ARG_KARUTA_QUIZ_ID, karutaQuizId)
            putInt(ARG_KAMI_NO_KU_STYLE, kamiNoKuStyle.ordinal)
            putInt(ARG_SHIMO_NO_KU_STYLE, shimoNoKuStyle.ordinal)
            putInt(ARG_ANIMATION_SPEED, animationSpeed.ordinal)
        }
    }
}
